server_management/management/commands/update.py

```diff
@@ -208,85 +40,95 @@ def handle(self, *args, **options):
                 with settings(warn_only=True):
-                    result = run('test -x /usr/bin/pypy')
+                    pypy = run('test -x /usr/bin/pypy')
 
-                if result.return_code == 0:
-                    sudo('virtualenv -p /usr/bin/pypy {}'.format(venv))
+                if pypy.return_code == 0:
+                    sudo(f'virtualenv -p /usr/bin/pypy {new_venv}')
                 else:
-                    sudo('virtualenv -p python{} {}'.format(python_version, venv))
-
-                with virtualenv(venv):
-                    with shell_env(DJANGO_SETTINGS_MODULE=settings_module):
-                        sudo('[[ -e requirements.txt ]] && pip install -qr requirements.txt')
+                    sudo(f'virtualenv -p python{python_version} {new_venv}')
 
-            with virtualenv(venv):
-                with shell_env(DJANGO_SETTINGS_MODULE=settings_module):
-                    sudo('pip install -q gunicorn')
+                with virtualenv(new_venv), shell_env(DJANGO_SETTINGS_MODULE=settings_module):
+                    sudo('[[ -e requirements.txt ]] && pip install -r requirements.txt')
+                    sudo('pip install gunicorn')
 
-                    if remote['server'].get('build_system', 'npm') == 'npm':
-                        sudo('. ~/.nvm/nvm.sh && yarn', shell='/bin/bash')
-                        sudo('. ~/.nvm/nvm.sh && yarn run build', shell='/bin/bash')
+            # Things which need to happen regardless of whether there was a venv already.
+            with virtualenv(new_venv), shell_env(DJANGO_SETTINGS_MODULE=settings_module):
+                if remote['server'].get('build_system', 'npm') == 'npm':
+                    sudo('. ~/.nvm/nvm.sh && yarn', shell='/bin/bash')
+                    sudo('. ~/.nvm/nvm.sh && yarn run build', shell='/bin/bash')
 
-                    sudo('python manage.py collectstatic --noinput')
+                sudo('python manage.py collectstatic --noinput')
 
-                    requirements = sudo('pip freeze')
-                    compressor = False
-                    watson = False
-                    for line in requirements.split('\n'):
-                        if line.startswith('django-compressor'):
-                            compressor = True
-                        if line.startswith('django-watson'):
-                            watson = True
+                requirements = sudo('pip freeze')
+                compressor = False
+                watson = False
+                for line in requirements.split('\n'):
+                    if line.startswith('django-compressor'):
+                        compressor = True
+                    if line.startswith('django-watson'):
+                        watson = True
 
-                    if not compressor:
-                        sudo('python manage.py compileassets')
+                if not compressor:
+                    sudo('python manage.py compileassets')
```

Comment:
Is this condition ever going to be reached? I am not sure if any of our `server-management`-using projects use `django-optimizations`, and almost certainly not our Py3 projects (bearing in mind py2 support was removed in this PR).

Reply:
That's a fair point. Can probably be removed.

Reply:
Actually, this isn't `django-optimizations`, it's `django-compressor` which we definitely still use. Not sure what `compileassets` does, but it was probably added for a reason.

Reply:
Are you sure? I was going by [this source file](https://pypkg.com/pypi/django-optimizations/f/src/optimizations/management/commands/compileassets.py) in `django-optimizations` and `manage.py compileassets` not being a known command in a stock project.

Reply:
Apparently I can't read the word 'not'

Reply:
Which word? :D